### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Charpoly): the universal characteristic polynomial

### PR DESCRIPTION
**Pull request by jcommelin**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+-/
+import Mathlib.Data.MvPolynomial.Equiv
+import Mathlib.Data.MvPolynomial.Funext
+import Mathlib.LinearAlgebra.Matrix.Charpoly.Coeff
+import Mathlib.RingTheory.MvPolynomial.Homogeneous
+
+/-!
+# The universal characteristic polynomial
+
+In this file we define the universal characteristic polynomial.
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Charpoly/Univ.lean:14]: Can you add a reference (or a very short math explanation) about the universal characteristic polynomial? It doesn't seem (at least to me) completely standard terminology, but maybe it is just my fault.

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+namespace Matrix.charpoly
+
+variable {R : Type*} (n : Type*) [CommRing R] [Fintype n] [DecidableEq n]
+
+/-- The universal characteristic polynomial for `n × n`-matrices.-/
+noncomputable
+def univ : Polynomial (MvPolynomial (n × n) ℤ) :=
+  charpoly <| Matrix.of fun i j ↦ MvPolynomial.X (i, j)
+
+lemma univ_map (M : Matrix n n R) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Charpoly/Univ.lean:38]:
```suggestion
@[simp]
lemma univ_map (M : Matrix n n R) :
```

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  simp only [univ, charpoly, det_apply', map_sum, _root_.map_mul, map_prod]
+  apply Finset.sum_congr rfl
+  rintro i -
+  congr 1
+  · simp only [AlgHom.toRingHom_eq_coe, map_intCast]
+  · apply Finset.prod_congr rfl
+    rintro j -
+    by_cases h : i j = j <;> simp [h]
+
+lemma univ_coeff_aeval (M : Matrix n n R) (i : ℕ) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Charpoly/Univ.lean:50]:
```suggestion
@[simp]
lemma univ_coeff_aeval (M : Matrix n n R) (i : ℕ) :
```

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/LinearAlgebra/Charpoly/Univ.lean]: I added some docs.

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+Its `i`-th coefficient is a homogeneous polynomial of degree `n - i`,
+see `Matrix.charpoly.univ_coeff_isHomogeneous`.
+
+By evaluating the coefficients at the entries of a matrix `M`,
+one obtains the characteristic polynomial of `M`,
+see `Matrix.charpoly.univ_map`. -/
+noncomputable
+def univ : Polynomial (MvPolynomial (n × n) ℤ) :=
+  charpoly <| Matrix.of fun i j ↦ MvPolynomial.X (i, j)
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Charpoly/Univ.lean:50]: This is probably an argument for dropping this definition entirely and using `charpoly mvPolynomialX`, perhaps with some renaming; or at least, for using an `abbrev`

[the remark this answers is not in the document]
Comment (jcommelin) [reply Mathlib/LinearAlgebra/Charpoly/Univ.lean]: Thanks, that's nice. I think and `abbrev` is worth it.

File: Mathlib/Data/MvPolynomial/Equiv.lean [hunk 5]
```diff
@@ -519,6 +519,17 @@ theorem degreeOf_coeff_finSuccEquiv (p : MvPolynomial (Fin (n + 1)) R) (j : Fin
     (support_coeff_finSuccEquiv.1 hm)
 #align mv_polynomial.degree_of_coeff_fin_succ_equiv MvPolynomial.degreeOf_coeff_finSuccEquiv
 
+lemma finSuccEquiv_rename_finSuccEquiv (e : σ ≃ Fin n) (φ : MvPolynomial (Option σ) R) :
```
Comment (eric-wieser) [inline Mathlib/Data/MvPolynomial/Equiv.lean:522]: I think a docstring might be helpful to explain this statement

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+noncomputable
+abbrev univ : Polynomial (MvPolynomial (n × n) R) :=
+  charpoly <| mvPolynomialX n n R
+
+variable {R}
+
+open MvPolynomial RingHomClass in
+@[simp]
+lemma univ_map_eval₂Hom (M : n × n → S) :
+    (univ R n).map (eval₂Hom f M) = charpoly (Matrix.of M.curry) := by
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Charpoly/Univ.lean:60]: Is there a more general result hiding here about `(charpoly A).map f` where `f` is an AlgHom?

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean [hunk 7]
```diff
@@ -255,6 +255,24 @@ lemma finSuccEquiv_coeff_isHomogeneous {N : ℕ} {φ : MvPolynomial (Fin (N+1))
   simpa [Finset.sum_subset_zero_on_sdiff (g := d.cons i)
     (d.cons_support (y := i)) (by simp) (fun _ _ ↦ rfl), Finset.sum_insert aux, ← h] using hφ hd
 
+-- TODO: develop API for `optionEquivLeft` and get rid of the `[Fintype σ]` assumption
+lemma coeff_isHomogeneous_of_optionEquivLeft_symm
+    [Fintype σ] {p : Polynomial (MvPolynomial σ R)}
```
Comment (eric-wieser) [inline Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:260]: Can this be relaxed to `Finite` in the meantime? Or is it needed by the statement?

File: Mathlib/Data/MvPolynomial/Basic.lean [hunk 8]
```diff
@@ -255,6 +255,12 @@ theorem C_inj {σ : Type*} (R : Type*) [CommSemiring R] (r s : R) :
   (C_injective σ R).eq_iff
 #align mv_polynomial.C_inj MvPolynomial.C_inj
 
+instance nontrivial_of_nontrivial (σ : Type*) (R : Type*) [CommSemiring R] [Nontrivial R] :
```
Comment (eric-wieser) [inline Mathlib/Data/MvPolynomial/Basic.lean:258]: Do we have this instance for `MonoidAlgebra`? If so it should be inheritable with a one-liner.

File: Mathlib/Data/MvPolynomial/Basic.lean [hunk 9]
```diff
@@ -255,6 +255,12 @@ theorem C_inj {σ : Type*} (R : Type*) [CommSemiring R] (r s : R) :
...
+    Nontrivial (MvPolynomial σ R) :=
+  nontrivial_of_ne 0 1 <| by
```
Comment (eric-wieser) [inline Mathlib/Data/MvPolynomial/Basic.lean:260]: I would hope we have `Injective.nontrivial` or similar for proving this more succinctly

File: Mathlib/LinearAlgebra/Charpoly/Univ.lean [hunk 10]
```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Data.MvPolynomial.Equiv
+import Mathlib.Data.MvPolynomial.Funext
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Charpoly/Univ.lean:7]:
```suggestion
```

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Comment by sgouezel** [comment]
bors r-
Can you merge master?
bors d+

**Pull request closed without merging**